### PR TITLE
UPSTREAM: <carry>: hack out the oapi for restmapping resources when m…

### PR DESCRIPTION
**Pull request by deads2k**
…ore than one is present

If there is more than one kind or resource for a value (the CLI uses this for deciding what to do about deploymentconfigs as a for instance), anything in that list that is an openshift resource (I've only listed one to prove the concept), will be removed.

I think that referencing names from a file will still work since that should come back with one match.

@juanvallejo can you see if files containing oapi resources still work.
@soltysh @mfojtik @smarterclayton is this a thing we can live with? It will finally make the groupified resources have priority on the CLI.
@DirectXMan12 I really wish this had not worked.

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package meta
+
+import "k8s.io/apimachinery/pkg/runtime/schema"
+
+var oapiResources = map[schema.GroupVersionResource]bool{
+	{Resource: "deploymentconfig"}:  true,
+	{Resource: "deploymentconfigs"}: true,
+	{Resource: "dc"}:                true,
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go]: there's less than 100 of these

**Comment by enj** [comment]
Uh this is the only way to force our groupified resources to be used?

**Comment by deads2k** [comment]
> Uh this is the only way to force our groupified resources to be used?

Fully open to other pulls.  Priority is based on groupversion.  The `""/v1` group/version is always first because the core API should win naming conflicts.  Everything else is based on discovery order.  We have hack in discovery that returns `/oapi` resources when the group/version `""/v1` is requested because `/oapi` resources have `""/v1` as their group/version (see yaml).

We have a special case for 75 or so resources in `""/v1` that they need to be behind the groupified API in priority.  I think this is the simplest way to make it happen.  I'm open to reasonable-ish alternatives.

**Comment by enj** [comment]
I cannot think of a better solution off the top of my head (I expect any solution to be some variation on this hack in the end).

**Comment by mfojtik** [comment]
(flake was the router/prometheus )

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go (hunk 1 above)
Comment (mfojtik) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go]: can we make this pluggable in upstream, so you can register "ignored" resources like:

```golang
type predicateFunc func(gvk) bool
var IgnoredResources = map[schema.GroupVersionResource]predicateFunc{}

// then in origin we will have something like:

meta.IgnoredResources[{Resource: "deploymentconfig"}] = func(gvk) { len(gvk.Group) == 0 }

```

Comment (mfojtik) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go]: (just thinking loud to avoid carry patch ;-)

Comment (deads2k) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/meta/patch.go]: > can we make this pluggable in upstream, so you can register "ignored" resources like

I don't think there's a use-case upstream.

**Comment by deads2k** [comment]
now with our 75 or so resources

**Comment by enj** [comment]
/lgtm

sadness

**Pull request merged**